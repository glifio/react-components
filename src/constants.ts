/* GLIF */
export const GLIF_EMAIL = 'squad@glif.io'
export const GLIF_CODE = 'https://github.com/glifio'
export const GLIF_NODES =
  'https://lotus.filecoin.io/docs/developers/hosted-lotus/'
export const GLIF_DISCORD = 'https://discord.gg/B9ju5Eu4Rq'
export const GLIF_TWITTER = 'https://twitter.com/glifio'
export const GLIF_BLOG = 'https://blog.glif.io'

/* WALLET TYPES */
export const HD_WALLET = 'HD_WALLET'
export const SINGLE_KEY = 'SINGLE_KEY'

/* LOGIN OPTIONS */
export const LEDGER = 'LEDGER'
export const METAMASK = 'METAMASK'
export const IMPORT_MNEMONIC = 'IMPORT_MNEMONIC'
export const CREATE_MNEMONIC = 'CREATE_MNEMONIC'
export const IMPORT_SINGLE_KEY = 'IMPORT_SINGLE_KEY'

/* NETWORK VARS */
export const MAINNET = 'f'
export const TESTNET = 't'
export const MAINNET_PATH_CODE = 461
export const TESTNET_PATH_CODE = 1
