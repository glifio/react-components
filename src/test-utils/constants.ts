export const WALLET_ADDRESS = 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi'
export const WALLET_ID = 't0101'

export const WALLET_ADDRESS_2 = 't1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a'
export const WALLET_ID_2 = 't0102'

export const WALLET_ADDRESS_3 = 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'
export const WALLET_ID_3 = 't0103'
