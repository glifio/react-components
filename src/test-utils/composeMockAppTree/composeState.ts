import { FilecoinNumber } from '@glif/filecoin-number'

import createPath from '../../utils/createPath'
import { IMPORT_MNEMONIC, IMPORT_SINGLE_KEY } from '../../constants'
import { mockWalletProviderInstance } from '../../../__mocks__/@glif/filecoin-wallet-provider'
import {
  WALLET_ADDRESS,
  WALLET_ADDRESS_2,
  WALLET_ID,
  WALLET_ID_2
} from '../constants'
import { WalletProviderState } from '../../../src/services/WalletProvider/types'

export const composeWalletProviderState = (
  initialWalletProviderState: WalletProviderState,
  preset:
    | 'preOnboard'
    | 'postOnboard'
    | 'postOnboardLowBal'
    | 'postOnboardWithError'
    | 'selectedOtherWallet'
) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            address: WALLET_ADDRESS,
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardLowBal': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            address: WALLET_ADDRESS,
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('.000001', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            address: WALLET_ADDRESS,
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY,
        error: 'ERROR'
      })
    }
    case 'selectedOtherWallet': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            address: WALLET_ADDRESS,
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          },
          {
            address: WALLET_ADDRESS_2,
            robust: WALLET_ADDRESS_2,
            id: WALLET_ID_2,
            balance: new FilecoinNumber('5', 'fil'),
            path: createPath(1, 1)
          }
        ],
        selectedWalletIdx: 1,
        loginOption: IMPORT_MNEMONIC
      })
    }
    default:
      return initialWalletProviderState
  }
}
