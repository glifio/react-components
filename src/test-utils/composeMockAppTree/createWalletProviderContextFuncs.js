import { FilecoinNumber } from '@glif/filecoin-number'
import { TESTNET_PATH_CODE } from '../../constants'
import createPath from '../../utils/createPath'

import {
  setError,
  resetLedgerState,
  resetState,
  walletList,
  updateBalance
} from '../../../src/services/WalletProvider/state'
import { mockWalletProviderInstance } from '../../../__mocks__/@glif/filecoin-wallet-provider'
import { WALLET_ADDRESS_3, WALLET_ID_3 } from '../constants'

export const mockFetchDefaultWallet = jest.fn().mockImplementation(() => ({
  balance: new FilecoinNumber('1', 'fil'),
  address: WALLET_ADDRESS_3,
  robust: WALLET_ADDRESS_3,
  id: WALLET_ID_3,
  path: createPath(TESTNET_PATH_CODE, 0)
}))

export default (walletProviderDispatch, state) => ({
  fetchDefaultWallet: mockFetchDefaultWallet,
  walletList: jest
    .fn()
    .mockImplementation(wallets => walletProviderDispatch(walletList(wallets))),
  setWalletError: jest
    .fn()
    .mockImplementation(errorMessage =>
      walletProviderDispatch(setError(errorMessage))
    ),
  connectLedger: jest.fn().mockImplementation(() => mockWalletProviderInstance),
  connectMetaMask: jest.fn().mockImplementation(() => {
    walletProviderDispatch({ type: 'METAMASK_CONFIGURED_SUCCESS' })
    return mockWalletProviderInstance
  }),
  getProvider: jest.fn().mockImplementation(() => mockWalletProviderInstance),
  resetLedgerState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetLedgerState())
  }),
  resetState: jest.fn().mockImplementation(() => {
    walletProviderDispatch(resetState())
  }),
  updateBalance: jest.fn().mockImplementation((balance, index) => {
    walletProviderDispatch(updateBalance(balance, index))
  }),
  walletError: jest.fn().mockImplementation(() => state.error),
  resetWalletError: jest.fn().mockImplementation(() => {})
})
