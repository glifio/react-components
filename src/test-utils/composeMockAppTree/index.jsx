/* eslint-disable react/prop-types */
import React from 'react'
import { mockWalletProviderInstance } from '../../../__mocks__/@glif/filecoin-wallet-provider'
import WalletProviderWrapper from '../../services/WalletProvider'
import { initialState as walletProviderInitialState } from '../../services/WalletProvider/state'
import { composeWalletProviderState } from './composeState'
import { TestEnvironment } from '../TestEnvironment'

jest.mock('../../services/WalletProvider')

const Index = (statePreset = 'preOnboard', options = {}) => {
  // here you can pass a walletProviderInitialState and a preset to shape the store how you want it for testing
  const initialState = composeWalletProviderState(
    options?.walletProviderInitialState || walletProviderInitialState,
    statePreset
  )

  let walletProviderCache = { ...initialState }

  const cacheWalletProviderState = state => {
    walletProviderCache = { ...state }
    return <></>
  }

  const getWalletProviderState = () => walletProviderCache

  const Tree = ({ children }) => {
    return (
      <TestEnvironment>
        <WalletProviderWrapper
          options={options}
          statePreset={statePreset}
          getState={cacheWalletProviderState}
          initialState={initialState}
        >
          {children}
        </WalletProviderWrapper>
      </TestEnvironment>
    )
  }

  return {
    Tree,
    walletProvider: mockWalletProviderInstance,
    getWalletProviderState
  }
}

/**
 * This function is a wrapper that mocks everything the filecoin app needs for testing
 *
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query
 */

export default Index
