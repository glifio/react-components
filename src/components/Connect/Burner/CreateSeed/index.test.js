import { render, screen, act, fireEvent } from '@testing-library/react'
import composeMockAppTree from '../../../../test-utils/composeMockAppTree'

import { CreateSeed } from '.'
import { TESTNET_PATH_CODE } from '../../../../constants'
import { mockFetchDefaultWallet } from '../../../../test-utils/composeMockAppTree/createWalletProviderContextFuncs'
import createPath from '../../../../utils/createPath'

const backSpy = jest.fn()
const nextSpy = jest.fn()

describe('Create seed phrase configuration', () => {
  beforeAll(() => {
    global.URL.createObjectURL = jest.fn()
  })

  test('it renders step 1 correctly', () => {
    const { Tree } = composeMockAppTree('preOnboard')

    render(
      <Tree>
        <CreateSeed initialWalkthroughStep={1} next={nextSpy} back={backSpy} />
      </Tree>
    )

    expect(screen.getByText(/Step 1/)).toBeInTheDocument()
    expect(screen.getByText(/Copy/)).toBeInTheDocument()
    expect(screen.getByText(/Download/)).toBeInTheDocument()
    expect(screen.getByText(/Write down your seed phrase/)).toBeInTheDocument()
    expect(screen.getByText(/I've recorded my seed phrase/)).toBeInTheDocument()
  })

  test('it renders step 2 correctly', () => {
    const { Tree } = composeMockAppTree('preOnboard')

    render(
      <Tree>
        <CreateSeed initialWalkthroughStep={2} next={nextSpy} back={backSpy} />
      </Tree>
    )

    expect(
      screen.getByText(/Add the correct words to the empty inputs/)
    ).toBeInTheDocument()
    expect(screen.getByText(/Step 2/)).toBeInTheDocument()
  })

  test('it renders step 3 correctly', () => {
    const { Tree } = composeMockAppTree('preOnboard')

    render(
      <Tree>
        <CreateSeed initialWalkthroughStep={3} next={nextSpy} back={backSpy} />
      </Tree>
    )

    expect(
      screen.getByText(/Success! Please click 'Next' to access your wallet./)
    ).toBeInTheDocument()
    expect(screen.getByText(/Step 3/)).toBeInTheDocument()
  })

  test('it sends the user to wallet view, with a wallet in state upon successful config', async () => {
    const { Tree, getWalletProviderState } = composeMockAppTree('preOnboard')

    const { container } = render(
      <Tree>
        <CreateSeed initialWalkthroughStep={3} next={nextSpy} back={backSpy} />
      </Tree>
    )

    await act(async () => {
      fireEvent.click(screen.getByText('Next'))
    })
    expect(container.firstChild).toMatchSnapshot()
    expect(mockFetchDefaultWallet).toHaveBeenCalled()
    expect(nextSpy).toHaveBeenCalled()
    const wallet = getWalletProviderState().wallets[0]
    expect(wallet.address).toBeTruthy()
    expect(wallet.path).toBe(createPath(TESTNET_PATH_CODE, 0))
  })
})
