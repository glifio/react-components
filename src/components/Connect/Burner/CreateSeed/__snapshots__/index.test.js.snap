// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create seed phrase configuration it renders the warning correctly 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  min-width: 0;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

.c2 {
  min-width: 0;
  box-sizing: border-box;
  border: 1px solid;
  border-radius: 4px;
  border-width: 1px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  max-width: 560px;
  min-height: 300px;
  overflow: hidden;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c3 {
  min-width: 0;
  box-sizing: border-box;
  padding: 16px;
  background-color: #FF9E80;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  min-width: 0;
  box-sizing: border-box;
  border-width: 0.1875rem;
  border-radius: 8px;
  border-style: solid;
  background-color: #FF9E80;
  color: #661800;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 48px;
  height: 48px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  min-width: 0;
  box-sizing: border-box;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 24px;
  padding-bottom: 24px;
}

.c10 {
  min-width: 0;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c11 {
  cursor: pointer;
  border: 1px solid #262626;
  background-color: transparent;
  border-color: #262626;
  color: #262626;
  font-size: 1.125rem;
  -webkit-transition: 0.18s ease-in-out;
  transition: 0.18s ease-in-out;
  border-radius: 4px;
  margin-top: 32px;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  height: 48px;
  border-width: 1px;
  border-radius: 4px;
}

.c11:hover {
  opacity: 0.8;
}

.c12 {
  cursor: pointer;
  border: 1px solid #1AD08F;
  background-color: #1AD08F;
  border-color: #1AD08F;
  color: #08442F;
  font-size: 1.125rem;
  -webkit-transition: 0.18s ease-in-out;
  transition: 0.18s ease-in-out;
  border-radius: 4px;
  margin-top: 32px;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  height: 48px;
  border-width: 1px;
  border-radius: 4px;
}

.c12:hover {
  opacity: 0.8;
}

.c6 {
  color: #661800;
  font-size: 1.5rem;
  font-weight: 400;
  line-height: 1.2;
  font-family: RT-Alias-Grotesk;
  margin: 0;
}

.c8 {
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.4;
  font-family: RT-Alias-Grotesk;
}

.c5 {
  font-family: "RT-Alias-Medium","system-ui","Segoe UI","Roboto",Helvetica;
  font-weight: 700;
  font-size: 1.5rem;
}

.c9 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.18s ease-in-out;
  transition: 0.18s ease-in-out;
  border-bottom: 2px solid #5E26FF00;
  color: #5E26FF;
  font-size: 1.25rem;
}

.c9:hover {
  border-bottom: 2px solid #5E26FF;
}

<div
  class="c0"
  display="flex"
>
  <div
    class="c1"
    display="flex"
    width="100%"
  >
    <div
      class="c2"
      display="flex"
      overflow="hidden"
      width="100%"
    >
      <div
        class="c3"
        display="flex"
      >
        <div
          class="c4"
          color="status.warning.foreground"
          display="flex"
          size="6"
        >
          <h3
            class="c5"
          >
            Wn
          </h3>
        </div>
        <h2
          class="c6"
          color="status.warning.foreground"
          font-family="RT-Alias-Grotesk"
          font-size="4"
          font-weight="400"
        >
          Warning
        </h2>
      </div>
      <div
        class="c7"
      >
        <p
          class="c8"
          font-family="RT-Alias-Grotesk"
          font-size="2"
          font-weight="400"
        >
          We do not recommend you use this burner wallet to hold or transact significant sums of Filecoin. For significant sums, Glif Sender should be used with a non-burner wallet.
        </p>
        <a
          class="c9"
          color="core.primary"
          font-size="3"
          href="https://coinsutra.com/security-risks-bitcoin-wallets/"
          rel="noopener"
          target="_blank"
        >
          Why isn't it secure?
        </a>
      </div>
    </div>
    <div
      class="c10"
      display="flex"
    >
      <button
        class="c11"
        font-size="3"
        height="6"
        type="button"
      >
        Back
      </button>
      <button
        class="c12"
        font-size="3"
        height="6"
        type="button"
      >
        I Understand
      </button>
    </div>
  </div>
</div>
`;

exports[`Create seed phrase configuration it sends the user to wallet view, with a wallet in state upon successful config 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1;
  font-family: RT-Alias-Grotesk;
  margin: 0;
  margin-top: 16px;
}

.c1 {
  display: inline-block;
  height: 48px;
  width: 48px;
  text-align: left;
  background-color: var(--purple-medium);
}

.c2 {
  display: inline-block;
  height: 24px;
  width: 24px;
  background-color: var(--purple-dark);
  -webkit-animation-name: bvGDKI;
  animation-name: bvGDKI;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

@media screen and (min-width:40em) {

}

@media screen and (min-width:40em) {

}

@media screen and (min-width:40em) {

}

<div
  class="c0"
  display="flex"
  height="100%"
  width="100%"
>
  <span
    class="c1"
  >
    <span
      class="c2"
    />
  </span>
  <h4
    class="c3"
    font-family="RT-Alias-Grotesk"
    font-size="2"
    font-weight="400"
  >
    Loading...
  </h4>
</div>
`;
